go-concourse/concourse/configs.go

```diff
@@ -79,46 +75,38 @@ func (team *team) CreateOrUpdatePipelineConfig(pipelineRef atc.PipelineRef, conf
 			"Content-Type":          {"application/x-yaml"},
 			atc.ConfigVersionHeader: {configVersion},
 		},
-	},
-		&response,
-	)
-
+	})
 	if err != nil {
-		if unexpectedResponseError, ok := err.(internal.UnexpectedResponseError); ok {
-			if unexpectedResponseError.StatusCode == http.StatusBadRequest {
-				var validationErr atc.SaveConfigResponse
-				err = json.Unmarshal([]byte(unexpectedResponseError.Body), &validationErr)
-				if err != nil {
-					return false, false, []ConfigWarning{}, err
-				}
-
-				return false, false, []ConfigWarning{}, InvalidConfigError{
-					Errors: validationErr.Errors,
-				}
-			}
-		}
-
 		return false, false, []ConfigWarning{}, err
 	}
 
-	configResponse := setConfigResponse{}
-	readCloser, ok := response.Result.(io.ReadCloser)
-	if !ok {
-		return false, false, []ConfigWarning{}, errors.New("Failed to assert type of response result")
-	}
-	defer readCloser.Close()
+	defer response.Body.Close()
+
+	if response.StatusCode < 200 || response.StatusCode >= 300 {
+		var validationErr atc.SaveConfigResponse
+		body, _ := ioutil.ReadAll(response.Body)
+		err = json.Unmarshal(body, &validationErr)
+		if err != nil {
+			return false, false, []ConfigWarning{}, internal.UnexpectedResponseError{
+				StatusCode: response.StatusCode,
+				Status:     response.Status,
+				Body:       string(body),
+			}
+		}
```

Comment:
Not your fault, but this a bit clunky, eh? We know enough about the config server to know that there are really two distinct scenarios here:
* you get a 400 status and an `atc.SaveConfigResponse`-shaped body
* you get a non-400 status and an unstructured body which [may](https://github.com/concourse/concourse/blob/e2831b531452357410a064dd4accce78fc6e87b5/atc/api/configserver/save.go#L113-L114) [be](https://github.com/concourse/concourse/blob/e2831b531452357410a064dd4accce78fc6e87b5/atc/api/configserver/save.go#L119-L120) [empty](https://github.com/concourse/concourse/blob/e2831b531452357410a064dd4accce78fc6e87b5/atc/api/configserver/save.go#L60-L61) or just contain an [error string](https://github.com/concourse/concourse/blob/e2831b531452357410a064dd4accce78fc6e87b5/atc/api/configserver/save.go#L126-L128)

How would you feel about casing on the specific status, adding a branch for `BadRequest`/`InvalidConfigError` and a branch for other non-200/`UnexpectedResponseError`?

Reply:
That sounds good. I think there are a couple more things it can return due to the api wrappas like this https://github.com/concourse/concourse/blame/master/atc/api/policychecker/handler.go#L45-L50. The specific thing I wanted here was an actual error message for 403s since OPA can provide useful 403s now. Would you be open to that being a specific new case and handing back a slightly more strongly typed Forbidden error in that case too?